sensor_array/analysis/sample_analysis.py

```diff
@@ -0,0 +1,1015 @@
+    del true_comp['N2_comp']
+    del true_comp['O2_comp']
+
+    return true_comp
+
+
+def calculate_all_arrays_list(mof_list, num_mofs):
+    mof_array_list = []
+    mof_array_list.extend(list(itertools.combinations(mof_list, num_mofs)))
+
+    return mof_array_list
+
+
+def import_prediction_data(prediction_results):
+    with open(prediction_results) as file:
+        reader = csv.reader(file, delimiter='\t')
+        reader_list = list(reader)
+        keys = reader_list[0]
+
+        prediction_results = []
+        for i in range(len(reader_list[1::])):
+            if int(i+0) % 4 == 0:
+                temp_dict = OrderedDict()
+                temp_dict[keys[0]] = int(reader_list[i+1][0])
+            elif int(i+3) % 4 == 0:
+                temp_dict[keys[1]] = str(reader_list[i+1][0])
+            elif int(i+2) % 4 == 0:
+                temp_dict[keys[2]] = ast.literal_eval(reader_list[i+1][0])
+            elif int(i+1) % 4 == 0:
+                temp_dict[keys[3]] = ast.literal_eval(reader_list[i+1][0])
+                prediction_results.extend([temp_dict])
+
+        return keys, prediction_results
+
+
+def plot_algorithm_progress_single_samples(gases, all_comp_sets, true_comp, cycle_nums, run_id, filepath):
+    colors = mpl.cm.get_cmap('RdBu')
+    color0 = colors(0.80)
+    color1 = colors(0.20)
+
+    for gas in gases:
+        comp_range = all_comp_sets[gas]
+        true_comp_as_array = true_comp[gas+'_comp']*np.ones(len(cycle_nums))
+        plt.figure(figsize=(4.5,4.5), dpi=300)
+        plt.rcParams['font.size'] = 12
+
+        if gas == 'CO2':
+            gas_for_title = '$Carbon$'+' '+'$Dioxide$'
+        else:
+            gas_for_title = '$' + gas[0].upper() + gas[1::] + '$'
+
+        plt.title(gas_for_title, fontsize=16)
+        plt.plot(cycle_nums, true_comp_as_array, '--', color='dimgrey', label='True Composition')
+        plt.xlabel('Cycle Number', fontsize=16)
+        xticks = [i for i in range(len(cycle_nums)) if i % 2==0]
+        plt.xticks(xticks, fontsize=12)
+        plt.ylabel('Mole Fraction', fontsize=16)
+        plt.yticks(fontsize=12)
+        plt.ticklabel_format(axis="y", style="sci", scilimits=(-1,3), useMathText=True)
+        for n in range(len(cycle_nums)):
+            if n == 0:
+                plt.plot([cycle_nums[n],cycle_nums[n]], comp_range[n], 'o-', color=color0, label='Predicted')
+            else:
+                plt.plot([cycle_nums[n],cycle_nums[n]], comp_range[n], 'o-', color=color0, label=None)
+        plt.legend(fontsize=12, bbox_to_anchor=(0.5,-0.15), loc='upper center', ncol=2)
+        plt.tight_layout(rect=(0,0.05,1,1))
+        plt.savefig(filepath+'breath_sample_prediction_algorithm_'+gas+'.png')
+        plt.close()
+
+
+def plot_predicted_vs_true_for_all_breath_samples(gases, gas_limits, predicted_comps, true_comps, filepath=None, sort_data=False, sort_by=None):
+    from matplotlib.ticker import AutoMinorLocator
+    from matplotlib.ticker import FixedLocator
+
```

Comment:
imports in the function? this implies the plots should be broken into a separate file